retrieval/impl/client.go

```diff
@@ -0,0 +1,381 @@
+}
+
+// TODO: Update to match spec for V0 epic
+// https://github.com/filecoin-project/go-retrieval-market-project/issues/8
+func (c *client) Query(ctx context.Context, p retrievalmarket.RetrievalPeer, pieceCID []byte, params retrievalmarket.QueryParams) (retrievalmarket.QueryResponse, error) {
+	cid, err := cid.Cast(pieceCID)
+	if err != nil {
+		log.Warn(err)
+		return retrievalmarket.QueryResponseUndefined, err
+	}
+
+	s, err := c.h.NewStream(ctx, p.ID, retrievalmarket.QueryProtocolID)
+	if err != nil {
+		log.Warn(err)
+		return retrievalmarket.QueryResponseUndefined, err
+	}
+	defer s.Close()
+
+	err = cborutil.WriteCborRPC(s, &OldQuery{
+		Piece: cid,
+	})
+	if err != nil {
+		log.Warn(err)
+		return retrievalmarket.QueryResponseUndefined, err
+	}
+
+	var oldResp OldQueryResponse
+	if err := oldResp.UnmarshalCBOR(s); err != nil {
+		log.Warn(err)
+		return retrievalmarket.QueryResponseUndefined, err
+	}
+
+	resp := retrievalmarket.QueryResponse{
+		Status:          retrievalmarket.QueryResponseStatus(oldResp.Status),
+		Size:            oldResp.Size,
+		MinPricePerByte: types.BigDiv(oldResp.MinPrice, types.NewInt(oldResp.Size)),
+	}
+	return resp, nil
+}
+
+// TODO: Update to match spec for V0 Epic:
+// https://github.com/filecoin-project/go-retrieval-market-project/issues/9
+func (c *client) Retrieve(ctx context.Context, pieceCID []byte, params retrievalmarket.Params, totalFunds types.BigInt, miner peer.ID, clientWallet retrievalmarket.Address, minerWallet retrievalmarket.Address) retrievalmarket.DealID {
+	/* The implementation of this function is just wrapper for the old code which retrieves UnixFS pieces
+	-- it will be replaced when we do the V0 implementation of the module */
+	c.nextDealLk.Lock()
+	c.nextDealID++
+	dealID := c.nextDealID
+	c.nextDealLk.Unlock()
+
+	dealState := retrievalmarket.ClientDealState{
+		DealProposal: retrievalmarket.DealProposal{
+			PieceCID: pieceCID,
+			ID:       dealID,
+			Params:   params,
+		},
+		Status: retrievalmarket.DealStatusFailed,
+		Sender: miner,
+	}
+
+	go func() {
+		evt := retrievalmarket.ClientEventError
+		converted, err := cid.Cast(pieceCID)
+
+		if err == nil {
+			err = c.retrieveUnixfs(ctx, converted, types.BigDiv(totalFunds, params.PricePerByte).Uint64(), totalFunds, miner, clientWallet, minerWallet)
+			if err == nil {
+				evt = retrievalmarket.ClientEventComplete
+				dealState.Status = retrievalmarket.DealStatusCompleted
+			}
+		}
+
+		c.notifySubscribers(evt, dealState)
+	}()
+
+	return dealID
+}
+
+// unsubscribeAt returns a function that removes an item from the subscribers list by comparing
+// their reflect.ValueOf before pulling the item out of the slice.  Does not preserve order.
+// Subsequent, repeated calls to the func with the same Subscriber are a no-op.
+func (c *client) unsubscribeAt(sub retrievalmarket.ClientSubscriber) retrievalmarket.Unsubscribe {
+	return func() {
+		curLen := len(c.subscribers)
+		for i, el := range c.subscribers {
```

Comment:
This feels like it should have a lock (and it should also be used in `notifySubscribers` / `SubscribeToEvents`)

Reply:
you are right. will add.